src/MongoDB.Driver/Encryption/MongocryptdFactory.cs

```diff
@@ -0,0 +1,177 @@
+﻿/* Copyright 2019-present MongoDB Inc.
+*
+* Licensed under the Apache License, Version 2.0 (the "License");
+* you may not use this file except in compliance with the License.
+* You may obtain a copy of the License at
+*
+* http://www.apache.org/licenses/LICENSE-2.0
+*
+* Unless required by applicable law or agreed to in writing, software
+* distributed under the License is distributed on an "AS IS" BASIS,
+* WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+* See the License for the specific language governing permissions and
+* limitations under the License.
+*/
+
+using System;
+using System.Collections;
+using System.Collections.Generic;
+using System.Diagnostics;
+using System.IO;
+using System.Linq;
+
+namespace MongoDB.Driver.Encryption
+{
+    internal class MongocryptdFactory
+    {
+        #region static
+        private static string[] __supportedExtraOptionKeys=
+        {
+            "mongocryptdURI",
+            "mongocryptdBypassSpawn",
+            "mongocryptdSpawnPath",
+            "mongocryptdSpawnArgs"
+        };
+        #endregion
+
+        private readonly IReadOnlyDictionary<string, object> _extraOptions;
+
+        public MongocryptdFactory(IReadOnlyDictionary<string, object> extraOptions)
+        {
+            _extraOptions = extraOptions ?? new Dictionary<string, object>();
+            EnsureThatExtraOptionsAreValid(extraOptions);
+        }
+
+        // public methods
+        public MongoClient CreateMongocryptdClient()
+        {
+            var connectionString = CreateMongocryptdConnectionString();
+            var clientSettings = MongoClientSettings.FromConnectionString(connectionString);
+            clientSettings.ServerSelectionTimeout = TimeSpan.FromMilliseconds(1000);
+            return new MongoClient(clientSettings);
+        }
+
+        public void SpawnMongocryptdProcessIfRequired()
+        {
+            if (ShouldMongocryptdBeSpawned(out var path, out var args))
+            {
+                StartProcess(path, args);
+            }
+        }
+
+        // private methods
+        private string CreateMongocryptdConnectionString()
+        {
+            if (_extraOptions.TryGetValue("mongocryptdURI", out var connectionString))
+            {
+                return (string)connectionString;
+            }
+            else
+            {
+                return "mongodb://localhost:27020";
+            }
+        }
+
+        private void EnsureThatExtraOptionsAreValid(IReadOnlyDictionary<string, object> extraOptions)
+        {
+            foreach (var extraOption in extraOptions)
+            {
+                if (!__supportedExtraOptionKeys.Contains(extraOption.Key))
+                {
+                    throw new ArgumentException($"Invalid extra option key: {extraOption.Key}.");
+                }
+            }
+        }
+
+        private bool IsMongoCryptBypassSpawnValid(object objectValue)
```

Comment:
Rename this method to `CastMongoCryptdBypassSpawnToBool`.

This method returns the actual value, not an indication of whether it is valid or not. If it is not valid an exception is thrown.